adding CONTRIBUTING.md and CONTRIBUTORS.md (#157)
VulcanClimateModeling/fv3core: CONTRIBUTING.rst

============
Contributing
============

FV3core is being actively developed by Vulcan Inc. Please contact us if there is interest
in making contributions. Credit will be given to contributors by adding their names
to the ``AUTHORS.rst`` file.

Development guidelines
----------------------

> Linting
> ~~~~~~~~~~

Reviewer: ```suggestion
Linting
~~~~~~~
```

The linter will become more strict over time, but you should always be able to correct (or

> at minimum learn what is incorrect) your code using `make lint`

Reviewer: ```suggestion
at minimum learn what is incorrect) your code using `make lint`.
```

> We manage the list of syntax requirements using `pre-commit <https://pre-commit.com/>`__

Reviewer: ```suggestion
We manage the list of syntax requirements using `pre-commit <https://pre-commit.com/>`__. This:
```

   - runs formatting and compliance checks for you

>    - required to pass to merge a PR

Reviewer: ```suggestion
   - is required to pass to merge a PR
```

> This initially was a call to 'black' `Black
> <https://github.com/ambv/black>`__ which reformats your code to a set of accepted

Reviewer: ```suggestion
This initially was a call to `Black
<https://github.com/ambv/black>`__ which reformats your code to a set of accepted
```

.pre-commit-config.yaml file.

When we add Flake 8 to pre-commit, these are the rules that are most important to us to
consider:

Allow (ignore rule):

- W503 line break before binary operator
    - We should choose whether to ignore W503 or W504, and only ignore one
- E302 whitespace before ‘:'
    - Needs to be ignored to be consistent with black

Add in a future PR but not immediately:

- F841 local variable is assigned to but never used
    - Probably have to ignore because that’s how stencils give outputs
    - Can avoid if all stencil assignments use square brackets on the left
        - out[0, 0, 0] or out[idx]/out[curr]/out[something]

Keep the rule:

- W504 line break after binary operator
    - ignoring W503 instead
- F401 imported but unused
    - Good for understanding what a file’s dependencies are
- F821 undefined name
    - To fix requires importing names explicitly
    - Useful because it catches real bugs
- E501 line too long (> 88 characters)
    - Should be included, these are lines black can’t handle automatically
- E265 block comment should start with “# “ and E262 inline comment should start with “# “

> File structure/ Conventions
> ~~~~~~~~~~

Reviewer: ```suggestion
File Structure / Conventions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

   defining the translation of variables from Fortran to Python. Exceptions exist in cases
   where topical and logical grouping allowed for code reuse. As refactors optimize the
   model, these units may be merged to occupy the same files and even methods/stencils, but
   the units should still be tested separately.
 - The core has most of its calculations happening in GT4py stencils, but there are still
   several instances of operations happening in Python directly, which will need to be
   replaced with GT4py code for optimal performance.
 - The 'units' fv_dynamics and fv_subgridz can be called by fv3gfs-wrapper to run the
   FV3core model using Pythong wrapped fortran for code not ported to GT4py at the moment.
 - The namelist and grid are global variables defined in fv3core/_config.py
   - The namelist is 'flattened' so that the grouping name of the option is not required
     to access the data (we may want to change this)
   - The grid variables are mostly 2d variables that have been replicated in the z
     dimension as gt4py storages for use in stencils
   - The grid object also contains domain and layout information relevant to the current
     rank being operated on.

>  - utility functions at fv3core/utils include:

Reviewer: ```suggestion
 - Utility functions at fv3core/utils include:
```

      - grid:
	 - A Grid class definition that provides information about the grid layout,
           current tile informationm access to grid variables used globally, and
           convenience methods related to tile indexing, origins and domains commonly used
         - A grid is defined for each MPI rank (minimum 6 ranks, 1 for each tile face of
           the cubed sphere grid represnting the whole Earth)
	 - Also provides functionality for generating a Quantity object (used to interface
           with the fv3gfs-wrapper, that allows us to run the full model, not just the
           dynamical core)
      - corners: port of corner calculations, initially direct Python calculations, being
        replaced with GT4py gtscript functions as the GT4py regions feature is implemented
      - mpi: a wrapper for importing mpi4py when available
      - global_constants.py: constants for use throughout the model
      - typing.py: Clean names for common types we use in the model. This is new and
        hasn't been adopted throughout the model yet, but will eventually be our
        standard. A shorthand 'sd' has been used in the intial version.

>  - The 'tests' currently includes a framework for translating fields serialized (using

Reviewer: ```suggestion
 - `tests/` currently includes a framework for translating fields serialized (using
```

   Serialbox from GridTools) from a Fortran run into gt4py storages that can be inputs to
   fv3core unit computations, and compares the results of the ported code to serialized
   data following a unit computation.

>  - docker : provides Dockerfiles for building a repeatable environment in which to run the

Reviewer: ```suggestion
 - `docker/` provides Dockerfiles for building a repeatable environment in which to run the
```

   core

>  - external: a folder for submoduled repos that provide essential functionality

Reviewer: ```suggestion
 - `external/`: a directory for submoduled repos that provide essential functionality
```

 - The build system uses Makefiles following the convention of other repos within
   VulcanClimateModeling

> Model Interface
> ~~~~~

Reviewer: ```suggestion
Model Interface
~~~~~~~~~~~~~~~
```

- More descriptive types on stencil definitions (typing.py)
- Docstrings on outward facing python functions: describe what methods are doing, describe
the intent (in, out, inout) of the function arguments
e.g. vorticitytransport_cgrid:
"""Update the C-Grid zonal and meridional velocity fields.

    Args: uc: x-velocity on C-grid (inout) vc: y-velocity on C-grid (inout) vort_c:
        Vorticity on C-grid (inout) ke_c: kinetic energy on C-grid (inout) v: y-velocit on
        D-grid (inout) u: x-velocity on D-grid (inout) dt2: timestep (input) """

Python functions (should mostly be light wrappers calling gt4py stencils, though currently
exceptions exist where python code does computations on data fields):
Original convention:
def compute(var1, var2, var3, param1, param2, param3):
Order of arguments did not matter
too much, but generally follows the convention of listing 3d fields first, followed by
parameters.

New convention: make use of typing.py to specify fields, also typehint any function
outputs. For example:
def compute(var1: FloatField, var2:IntField, var3: BoolField,
param1: float_type, param2: int_type, param3: bool_type):
"""

> Describe what is being computed by this method
> Args:

Reviewer: ```suggestion
Describe what is being computed by this method

Args:
```

>   var1: description of data field (inout)
>   var2: description of int field (in)

Reviewer: I think @mcgibbon proposed putting these at the beginning after all
```suggestion
  var1 (inout): description of data field
  var2 (in): description of int field
```

> etc

Reviewer: ```suggestion
  ...
```

> ============

Author: Thanks for pointing that out, yeah the formatting was all kinds of screwed up. Hopefully is improved now.

Contributing
============

FV3core is being actively developed by Vulcan Inc. Please contact us if there is interest
in making contributions. Credit will be given to contributors by adding their names

> to the ``AUTHORS.rst`` file.

Author: That sounds fine to me.

When we add Flake 8 to pre-commit, these are the rules that are most important to us to
consider:

Allow (ignore rule):

- W503 line break before binary operator
    - We should choose whether to ignore W503 or W504, and only ignore one
- E302 whitespace before ‘:'
    - Needs to be ignored to be consistent with black

Add in a future PR but not immediately:

- F841 local variable is assigned to but never used
    - Probably have to ignore because that’s how stencils give outputs
    - Can avoid if all stencil assignments use square brackets on the left
        - out[0, 0, 0] or out[idx]/out[curr]/out[something]

Keep the rule:

- W504 line break after binary operator
    - ignoring W503 instead
- F401 imported but unused
    - Good for understanding what a file’s dependencies are
- F821 undefined name
    - To fix requires importing names explicitly
    - Useful because it catches real bugs
- E501 line too long (> 88 characters)
    - Should be included, these are lines black can’t handle automatically
- E265 block comment should start with “# “ and E262 inline comment should start with “# “

File Structure / Conventions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

>  - The main functionality of the FV3 dynamical core, which has been ported from the

Author: I turned some into paragraphs and moved some of this into the README instead.
